source/validate.h

```diff
@@ -84,91 +106,280 @@ enum class FunctionDecl {
+  uint32_t get_id() const { return id_; }
+
+  /// Returns the predecessors of the BasicBlock
+  const std::vector<BasicBlock*>& get_predecessors() const {
+    return predecessors_;
+  }
+
+  /// Returns the predecessors of the BasicBlock
+  std::vector<BasicBlock*>& get_predecessors() { return predecessors_; }
+
+  /// Returns the successors of the BasicBlock
+  const std::vector<BasicBlock*>& get_successors() const { return successors_; }
+
+  /// Returns the successors of the BasicBlock
+  std::vector<BasicBlock*>& get_successors() { return successors_; }
+
+  /// Sets the immedate dominator of this basic block
+  ///
+  /// @param[in] dom_block The dominator block
+  void SetImmediateDominator(BasicBlock* dom_block);
+
+  /// Returns the immedate dominator of this basic block
+  BasicBlock* GetImmediateDominator();
+
+  /// Returns the immedate dominator of this basic block
+  const BasicBlock* GetImmediateDominator() const;
+
+  /// Sets the @p next BasicBlock as the successor of this BasicBlock
+  void RegisterSuccessor(BasicBlock& next);
+
+  /// Sets the @p next BasicBlock as the successor of this BasicBlock
+  void RegisterSuccessor(std::vector<BasicBlock*> next);
+
+  /// Returns true if the id of the BasicBlock matches
+  bool operator==(const BasicBlock& other) const { return other.id_ == id_; }
+
+  /// Returns true if the id of the BasicBlock matches
+  bool operator==(const uint32_t& id) const { return id == id_; }
+
+  /// @brief A BasicBlock dominator iterator class
+  ///
+  /// This iterator will iterate to the the immediate dominators
+  /// of the block
+  class DominatorIterator : public std::iterator<std::forward_iterator_tag, BasicBlock*> {
+   public:
+    /// @brief Constructs the end of dominator iterator
+    ///
+    /// This will create an iterator which will represent the element
+    /// before the root node of the dominator tree
+    DominatorIterator();
+
+    /// @brief Constructs an iterator for the given block which points
+    DominatorIterator(BasicBlock* block);
+
+    /// @brief Advances the iterator
+    DominatorIterator& operator++();
+
+    /// @brief Returns the current element
+    BasicBlock*& operator*();
+
+    friend bool operator==(const DominatorIterator& lhs,
+                           const DominatorIterator& rhs);
+
+   private:
+    BasicBlock* current_;
+  };
+
+  DominatorIterator dom_begin();
+
+  DominatorIterator dom_end();
+
+  friend std::ostream& operator<<(std::ostream& os, const BasicBlock& other);
+
+ private:
+  // Id of the BasicBlock
+  const uint32_t id_;
+
+  // Pointer to the immediate dominator of the BasicBlock
+  BasicBlock* immediate_dominator_;
+
+  // The set of predecessors of the BasicBlock
+  std::vector<BasicBlock*> predecessors_;
+
+  // The set of successors of the BasicBlock
+  std::vector<BasicBlock*> successors_;
+  ValidationState_t& module_;
+  Function* function_;
+};
+
+/// @brief Returns true if the iterators point to the same element
```

Comment:
Seems odd.  I'd expect it to return true if the arguments were the same iterator (including `dom_end()`, which presumably doesn't point to anything).  If I want to check the equality of the pointed-to elements, I can always dereference the iterators.